### DISCUSSION TITLE
DV for Encoded MKV files to X265

### DISCUSSION THREAD

**Discussion by PhantomICEMAN** [quietvoid/dovi_tool#311] [Q&A]
Hi i currently have files that are mkv's that have DV layer of 7.6 as an enhanced layer. I am currently trying to encode these to x265 to make them smaller using ffmpeg upon the completion i want to re add the dv meta data back in. I have tried using the RPU injection method however when i try and run the file in plex it says the file cannot be played. As i understand the injection method doesnt work with every player out there. I have tried to convert the enhanced layer back to profile 8  without any encoding and that works flawlessly but i cant get my head round how you can do it with a rencoded video file. I have seen this done on file i have in my library but cant figure out how. Any ideas?

**Comment by fkroener** [comment]
I think I might have related questions. I have a 4k dolby mkv, for which mpv complains it can't handle the EL, thus "Dolby Vision, Version 1.0, Profile 7.6, dvhe.07.06, BL+EL+RPU", colormatrix says: bt2020-ncl.

I can convert this into a BL_RPU.hevc, which turns the colormatrix to dolbyvision and mpv appears to like it better. But I'd like to have it 1080p instead and I don't think I can just use ffmpeg to reencode it with `-filter:v "crop=3840:2076:0:42,scale=1920:-1"`, can I?

Following this [reddit post](https://www.reddit.com/r/ffmpeg/comments/rv3sm5/downscaling_4k_hdr_10bit_to_1080p_but_maintaining/) I extracted the RPU.bin (with -c) and then ran
`ffmpeg -y -i <file.mkv> -map 0:v:0 -filter:v "crop=3840:2076:0:42,scale=1920:-1" -v error -stats -c:v libx265 -preset veryfast -pix_fmt yuv420p10le -profile:v main10 -vsync 0 -level 5.1 -max_muxing_queue_size 4096  -x265-params crf=18:vbv-bufsize=10800:vbv-maxrate=5400:selective-sao=0:no-sao=1:strong-intra-smoothing=0:rect=0:aq-mode=1:hdr-opt=1:dolby-vision-profile=8.1:dolby-vision-rpu='RPU.bin':dhdr10-opt=1:repeat-headers=1:colorprim=bt2020:range=limited:transfer=smpte2084:colormatrix=bt2020nc:range=limited:master-display="G(8500,39850)B(6550,2300)R(35400,14600)WP(15635,16450)L(10000000,1)":max-cll="0,0" -bsf:v hevc_mp4toannexb -f hevc '/tmp/1080p.hevc'`

which didn't seem to actually use the RPU at all, so afterwards I injected it once more on the 1080p.hevc.

Merging it into an mkv container I now see colormatrix bt2020-ncl again, although I expected it to be dolbyvision. mediainfo now states for the video stream `Dolby Vision, Version 1.0, Profile 7.6, dvhe.07.03, BL+EL+RPU, no metadata compression, Blu-ray compatible / SMPTE ST 2086, Version HDR10, HDR10 compatible`, I thought I expected it to be 8.1 - but mpv now happily plays it.

Oh, and converting this file now with -m 2 gets me the 8.1 profile.  

Any pointers on how I could better wield ffmpeg/x265 and dovi-tool to downscale and reencode 4k files?

**Comment by fkroener** [comment]
It appears setting dolbyvision=1 is enough with recent ffmpeg/libx265 versions for re-encoding and downscaling the video and while keeping the mkv as one with audio, subs and metadata.

Something like `-filter:v "crop=3840:1632:0:264,scale=1920:-1" -c:v libx265 -dolbyvision 1` together with vbv-bufsize of your liking (e.g. `-x265-params vbv-bufsize=11600:vbv-maxrate=5800`) should take care of carrying HDR10 and DoVi metadata over, without needing any of colorprim/transfer/colormatrix/master-display declared. Though adding `repeat-headers=1` removes a warning from libx265.

Interestingly mediainfo reports profile 8.1, while bt2020-ncl suggests it is actually 7.6. But mpv doesn't complain about not supporting the EL and I get to see "hevc: Multiple Dolby Vision RPUs found in one AU. Skipping previous." while seeking. Unsure if dynamic HDR is actually used though.

One `ffmpeg -i <file.mkv> -c:v copy -bsf:v hevc_mp4toannexb -f hevc - | dovi_tool -m 2 convert -d -` and `mkvmerge -o final.mkv BL_RPU.hevc -D <file.mkv>` gives the proper 8.1.

ffmpeg/libx265 doesn't handle converting the level 5 metadata (active area offsets) though, which means one needs `--crop` above. So there isn't too much won by the dovi metadata passthrough, besides not needing to declare all the hdr options.
